### PR TITLE
Implement HDFS for go-storage

### PR DESCRIPTION
**Pull request by bokket**
Linked issue #1
## Choose a well maintained HDFS library
Choose [hdfs](https://github.com/colinmarc/hdfs) to operate HDFS API
##  Implement all interface that Storager required
I've only implemented the basic implementation for now, and I will continue to expand and redesign the details
### Basic Operations

- [x] Create
- [x] Delete
- [x] List
- [x] Metadata
- [x] Read
- [x] Stat
- [x] Write

## Others
Pass integration tests

### CODE REVIEW

**Comment by bokket** [comment]
It need [setup-hdfs](https://github.com/beyondstorage/setup-hdfs) for action CI, I will do it tomorrow.

File: Makefile [hunk 1]
```diff
@@ -1,5 +1,7 @@
 SHELL := /bin/bash
 
+-include Makefile.env.example
```
Comment (Xuanwo) [inline Makefile]:
```suggestion
-include Makefile.env
```

File: go.mod [hunk 2]
```diff
@@ -2,4 +2,11 @@ module github.com/beyondstorage/go-service-hdfs
 
 go 1.15
 
-require github.com/beyondstorage/go-storage/v4 v4.4.0
+require (
+	github.com/beyondstorage/go-endpoint v1.0.2-0.20210719035101-60d3c8bfe7dd
```
Comment (Xuanwo) [inline go.mod]: https://github.com/beyondstorage/go-endpoint/issues/12

File: service.toml [hunk 3]
```diff
@@ -19,3 +23,35 @@ optional = ["object_mode"]
...
+
+[pairs.storage_features]
+type = "StorageFeatures"
+description = "set storage features"
+
+[pairs.default_storage_pairs]
+type = "DefaultStoragePairs"
+description = "set default pairs for storager actions"
+
+[infos.object.meta.name]
```
Comment (Xuanwo) [inline service.toml]: We don't need to store the object's name, `Path` is enough.

File: service.toml [hunk 4]
```diff
@@ -19,3 +23,35 @@ optional = ["object_mode"]
...
+type = "string"
+description = "base name of the file"
+
+[infos.object.meta.size]
```
Comment (Xuanwo) [inline service.toml]: Please use `content-length` instead.

File: service.toml [hunk 5]
```diff
@@ -19,3 +23,35 @@ optional = ["object_mode"]
...
+type = "int64"
+description = "length in bytes for regular files; system-dependent for others"
+
+[infos.object.meta.mode]
```
Comment (Xuanwo) [inline service.toml]: Please use `ObjectMode` instead.

File: service.toml [hunk 6]
```diff
@@ -19,3 +23,35 @@ optional = ["object_mode"]
...
+type = "uint32"
+description = "file mode bits"
+
+[infos.object.meta.ModTime]
```
Comment (Xuanwo) [inline service.toml]: Please use `LastModified` instead.

File: service.toml [hunk 7]
```diff
@@ -19,3 +23,35 @@ optional = ["object_mode"]
...
+type = "time.Time"
+description = "modification time"
+
+[infos.object.meta.IsDir]
```
Comment (Xuanwo) [inline service.toml]: Please use `ObjectMode` instead.

File: service.toml [hunk 8]
```diff
@@ -19,3 +23,35 @@ optional = ["object_mode"]
...
+type = "bool"
+description = "abbreviation for Mode().IsDir()"
+
+[infos.object.meta.Sys]
```
Comment (Xuanwo) [inline service.toml]: Please remove this meta and we will provide `SystemMetadata` instead.

File: storage.go [hunk 9]
```diff
@@ -2,35 +2,192 @@ package hdfs
 
 import (
 	"context"
+	"errors"
 	"io"
+	"os"
+	path2 "path"
```
Comment (Xuanwo) [inline storage.go]: Please rename the usage of `path` instead rename the import here.

File: storage.go [hunk 10]
```diff
@@ -2,35 +2,192 @@ package hdfs
...
 
+	"github.com/beyondstorage/go-storage/v4/pkg/iowrap"
+	"github.com/beyondstorage/go-storage/v4/services"
 	. "github.com/beyondstorage/go-storage/v4/types"
 )
 
+// ref: [hdfs#Client.Stat](https://pkg.go.dev/github.com/colinmarc/hdfs#Client.Stat)
+// Stat returns an os.FileInfo describing the named file or directory.
+// But FileInfo.Size don't return a correct write byte size
```
Comment (Xuanwo) [inline storage.go]: > But FileInfo.Size don't return a correct write byte size

Can you explain in more detail?

File: storage.go [hunk 11]
```diff
@@ -2,35 +2,192 @@ package hdfs
...
+	}
+	return nil
 }
 
 func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	if !opt.HasListMode || opt.ListMode.IsDir() {
+		nextFn := func(ctx context.Context, page *ObjectPage) error {
+			dir, err := s.hdfs.ReadDir(rp)
```
Comment (Xuanwo) [inline storage.go:37]: Please `ReadDir` by offset instead read all content at once. The dir could be very large.

File: storage.go [hunk 12]
```diff
@@ -2,35 +2,192 @@ package hdfs
...
+		return err
+	}
+	return nil
 }
 
 func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	if !opt.HasListMode || opt.ListMode.IsDir() {
+		nextFn := func(ctx context.Context, page *ObjectPage) error {
```
Comment (Xuanwo) [inline storage.go:36]: How about adding new functions as other services did? The logic for `list` is so complex.

File: storage.go [hunk 13]
```diff
@@ -2,35 +2,192 @@ package hdfs
...
+		o.Mode |= ModeDir
+	} else {
+		o.Mode |= ModeRead
+	}
+
+	o.SetContentLength(*Size)
+
+	var sm ObjectSystemMetadata
+
+	sm.Name = stat.Name()
```
Comment (Xuanwo) [inline storage.go]: See comments on `service.toml`.

`go-storage` vision is `Write once, run on every storage service.`, If we define different metadata for every service, we could reach this vision.

File: storage.go [hunk 14]
```diff
@@ -2,35 +2,192 @@ package hdfs
...
+	err = s.hdfs.MkdirAll(dir, 0666)
+	if err != nil {
+		return 0, err
+	}
+	file := dir + "/" + filename
+
+	var f io.Writer
+
+	_, err = s.hdfs.Stat(file)
+	if os.IsNotExist(err) {
```
Comment (Xuanwo) [inline storage.go]: How about check the file and remove it if exist, and then `create`?

File: tests/storage_test.go [hunk 15]
```diff
@@ -0,0 +1,15 @@
+package tests
+
+import (
+	tests "github.com/beyondstorage/go-integration-test/v4"
```
Comment (Xuanwo) [inline tests/storage_test.go]: import format.

File: tests/utils_test.go [hunk 16]
```diff
@@ -0,0 +1,24 @@
+package tests
+
+import (
+	hdfs "github.com/beyondstorage/go-service-hdfs"
+	"github.com/beyondstorage/go-storage/v4/pairs"
```
Comment (Xuanwo) [inline tests/utils_test.go:7]: ditto

File: utils.go [hunk 17]
```diff
@@ -1,27 +1,126 @@
...
+	workDir string
+
 	types.UnimplementedStorager
+	types.UnimplementedDirer
 }
 
 // String implements Storager.String
 func (s *Storage) String() string {
-	panic("implement me")
+	return fmt.Sprintf("Storager HDFS {WorkDir:%s}", s.workDir)
```
Comment (Xuanwo) [inline utils.go]:
```suggestion
	return fmt.Sprintf("Storager hdfs {WorkDir: %s}", s.workDir)
```

File: utils.go [hunk 18]
```diff
@@ -1,27 +1,126 @@
...
+		Storager: s,
+		Path:     path,
+	}
+}
+
+func (s *Storage) newObject(done bool) *types.Object {
+	return types.NewObject(s, done)
+}
+
+func (s *Storage) getFileName(path string) string {
```
Comment (Xuanwo) [inline utils.go]: How about use `path.Base()` instead?

File: go.mod [hunk 19]
```diff
@@ -2,4 +2,11 @@ module github.com/beyondstorage/go-service-hdfs
...
+	github.com/beyondstorage/go-integration-test/v4 v4.2.0
+	github.com/beyondstorage/go-storage/v4 v4.3.2
```
Comment (Xuanwo) [inline go.mod]:
```suggestion
	github.com/beyondstorage/go-storage/v4 v4.4.0
```

File: storage.go (hunk 10 above)
Comment (bokket) [reply storage.go]: I wrote a test and found that `hdfs.Stat() `returned the `size()` member size of 0, but the actual write size should exist

File: storage.go (hunk 14 above)
Comment (bokket) [reply storage.go]: Yes, this needs to be streamlined

File: storage.go (hunk 10 above)
Comment (Xuanwo) [reply storage.go]: Let's defer this discussion until we set up the test for hdfs.

File: storage.go (hunk 12 above)
Comment (bokket) [reply storage.go:36]: This may take some time for trial and error, can it be improved in the next pr?

Comment (Xuanwo) [reply storage.go:36]: LGTM, please start an issue to record.

File: storage.go (hunk 10 above)
Comment (bokket) [reply storage.go]: > I wrote a test and found that `hdfs.Stat() `returned the `size()` member size of 0, but the actual write size should exist

My mistake，it should be ensured that `file.close()` is released manually, instead `defer file.close()`

File: go.mod [hunk 20]
```diff
@@ -2,4 +2,11 @@ module github.com/beyondstorage/go-service-hdfs
 
 go 1.15
 
-require github.com/beyondstorage/go-storage/v4 v4.4.0
+require (
+	github.com/beyondstorage/go-endpoint v1.0.2-0.20210728080051-7a79fb7354dd // indirect
```
Comment (Xuanwo) [inline go.mod]:
```suggestion
	github.com/beyondstorage/go-endpoint v1.1.0
```

File: storage.go [hunk 21]
```diff
@@ -2,35 +2,152 @@ package hdfs
...
+		return 0, err
+	}
+	_, err = s.hdfs.Stat(rp)
+	if err == nil {
+		err = s.hdfs.Remove(rp)
+		if err != nil && errors.Is(err, os.ErrNotExist) {
+			// Omit `file not exist` error here
+			// ref: [GSP-46](https://github.com/beyondstorage/specs/blob/master/rfcs/46-idempotent-delete.md)
+			err = nil
+		}
```
Comment (Xuanwo) [inline storage.go:131]: We need to return the error if `Remove` meets other errors.

File: storage.go [hunk 22]
```diff
@@ -2,35 +2,152 @@ package hdfs
...
+	}
+	var f io.Writer
```
Comment (Xuanwo) [inline storage.go]: `FileWriter` has implemented `io.Writer` already, maybe we can remove the `f` here?

File: storage.go [hunk 23]
```diff
@@ -2,35 +2,152 @@ package hdfs
...
+	filewrite, err := s.hdfs.Create(rp)
+	if err != nil {
+		return 0, err
+	}
+	f = filewrite
```
Comment (Xuanwo) [inline storage.go]: How about add defer logic here:

```go
defer func() {
    err = filewrite.Close()
}
```

NOTE: the file could be kept open if we meet an error in `io.CopyN`.

File: storage.go [hunk 24]
```diff
@@ -2,35 +2,152 @@ package hdfs
...
+	if opt.HasIoCallback {
+		r = iowrap.CallbackReader(r, opt.IoCallback)
+	}
+
+	n, err = io.CopyN(f, r, size)
+	if err != nil {
+		return 0, err
+	}
+	filewrite.Flush()
```
Comment (Xuanwo) [inline storage.go]: As we will close the file, maybe we don't need to call `Flush` here?

**Review by Xuanwo — APPROVED** [review]

**Comment by Xuanwo** [comment]
Very good work, thanks for your contribution!

**Pull request merged**